File: boj/graph/16929.java
import java.util.*;

public class Main {

    static int N, M;
    static char[][] map;
    static boolean[][] visited;
    static int[][] dist;
    static final int[] toRow = {0, 0, 1, -1}, toCol = {1, -1, 0, 0};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        N = sc.nextInt();
        M = sc.nextInt();

        map = new char[N][M];

        for (int i = 0; i < N; i++) {
            String input = sc.next();
            for (int j = 0; j < M; j++) {
                map[i][j] = input.charAt(j);
            }
        }

        for (int i = 0; i < N - 1; i++) {
            for (int j = 0; j < M - 1; j++) {
                visited = new boolean[N][M];
                dist = new int[N][M];

                if (search(i, j, 0, map[i][j])) {
                    System.out.println("Yes");
                    return;
                }
            }
        }

        System.out.println("No");
    }

    static boolean search(int row, int col, int cnt, char value) {
        if (visited[row][col]) {
            return cnt - dist[row][col] >= 4;
        }

        visited[row][col] = true;
        dist[row][col] = cnt;

        for (int i = 0; i < 4; i++) {
            int nextRow = row + toRow[i];
            int nextCol = col + toCol[i];

            if (!isInRange(nextRow, nextCol)) continue;
            if (map[nextRow][nextCol] != value) continue;

            if (search(nextRow, nextCol, cnt + 1, value))
                return true;
        }

        return false;
    }

    static boolean isInRange(int r, int c) {
        if (r < 0 || c < 0 || r >= N || c >= M) return false;
        return true;
    }
}

File: boj/graph/16929.swift
import Foundation

let toward = [[0, 1], [0, -1], [1, 0], [-1, 0]]

var N = Int(), M = Int()
var map = [[String]]()
var visited = [[Bool]]()

struct Position: Equatable {
    var row: Int, col: Int
    
    init(_ row: Int, _ col: Int) {
        self.row = row
        self.col = col
    }
    
    static func == (lhs: Position, rhs: Position) -> Bool {
        return lhs.row == rhs.row && lhs.col == rhs.col
    }
}

func search(_ current: Position, _ length: Int, _ start: Position) {
    if current == start && length >= 4 {
        print("Yes")
        exit(0)
    }
    
    if visited[current.row][current.col] { return }
    visited[current.row][current.col] = true

    for i in 0..<toward.count {
        let nextRow = current.row + toward[i][0]
        let nextCol = current.col + toward[i][1]
        
        if nextRow < 0 || nextCol < 0 || nextRow >= N || nextCol >= M { continue }
        if map[nextRow][nextCol] != map[current.row][current.col] { continue }
        
        search(Position(nextRow, nextCol), length+1, start)
    }
    
    visited[current.row][current.col] = false
}

func main() {
    guard let input = readLine()?.split(separator: " ").compactMap({ Int(String($0)) }) else {
        print("error of reading input")
        return
    }
    
    N = input[0]
    M = input[1]
    
    map = [[String]](repeating: [String](repeating: String(), count: M), count: N)
    visited = [[Bool]](repeating: [Bool](repeating: false, count: M), count: N)
    
    for i in 0..<N {
        guard let row = readLine()?.compactMap({String($0)}) else {
            print("error of reading row")
            return
        }
        for j in 0..<M {
            map[i][j] = row[j]
        }
    }
    
    for i in 0..<N-1 {
        for j in 0..<M-1 {
            search(Position(i, j), 0, Position(i, j))
        }
    }
    print("No")
}


main()
